cmd/installer/cli/upgrade.go

```diff
@@ -0,0 +1,439 @@
+	// this does not return an error - it returns the previous umask
+	_ = syscall.Umask(0o022)
+
+	// Use the user-provided manager port if specified, otherwise use the existing runtime config value
+	if flags.managerPort != 0 {
+		// User provided a custom manager port
+		upgradeConfig.managerPort = flags.managerPort
+	} else {
+		// Use existing manager port from runtime config
+		upgradeConfig.managerPort = rc.ManagerPort()
+	}
+
+	eucfg, err := helpers.ParseEndUserConfig(flags.overrides)
+	if err != nil {
+		return fmt.Errorf("process overrides file: %w", err)
+	}
+	upgradeConfig.endUserConfig = eucfg
+
+	cv, err := getCurrentConfigValues()
+	if err != nil {
+		return fmt.Errorf("failed to get current config values: %w", err)
+	}
+	upgradeConfig.configValues = cv
+
+	return nil
+}
+
+func verifyAndPromptUpgrade(ctx context.Context, flags UpgradeCmdFlags, upgradeConfig upgradeConfig, prompt prompts.Prompt) error {
+	isAirgap := flags.airgapBundle != ""
+
+	err := verifyChannelRelease("upgrade", isAirgap, flags.assumeYes)
+	if err != nil {
+		return err
+	}
+
+	if upgradeConfig.airgapMetadata != nil && upgradeConfig.airgapMetadata.AirgapInfo != nil {
+		logrus.Debugf("checking airgap bundle matches binary")
+		if err := checkAirgapMatches(upgradeConfig.airgapMetadata.AirgapInfo); err != nil {
+			return err // we want the user to see the error message without a prefix
+		}
+	}
+
+	if !isAirgap {
+		if err := maybePromptForAppUpdate(ctx, prompt, upgradeConfig.license, flags.assumeYes); err != nil {
+			if errors.As(err, &ErrorNothingElseToAdd{}) {
+				return err
+			}
+			// If we get an error other than ErrorNothingElseToAdd, we warn and continue as this
+			// check is not critical.
+			logrus.Debugf("WARNING: Failed to check for newer app versions: %v", err)
+		}
+	}
+
+	if err := release.ValidateECConfig(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getCurrentConfigValues() (apitypes.AppConfigValues, error) {
+	// TODO: implement getting current config values from kots
+	return apitypes.AppConfigValues{}, nil
+}
+
+// getClusterID gets the cluster ID from the latest installation in the cluster
+func getClusterID(ctx context.Context, kcli client.Client) (string, error) {
+	installation, err := kubeutils.GetLatestInstallation(ctx, kcli)
+	if err != nil {
+		return "", err
+	}
+	if installation.Spec.ClusterID == "" {
+		return "", fmt.Errorf("existing installation has empty cluster ID")
+	}
+
+	return installation.Spec.ClusterID, nil
+}
+
+// readTLSConfig reads the TLS certificate from the kotsadm-tls secret
+func readTLSConfig(ctx context.Context, kcli client.Client) (apitypes.TLSConfig, error) {
+	var tlsConfig apitypes.TLSConfig
+
+	tlsSecret := &corev1.Secret{}
+	err := kcli.Get(ctx, client.ObjectKey{
+		Namespace: constants.KotsadmNamespace,
+		Name:      "kotsadm-tls",
+	}, tlsSecret)
+	if err != nil {
+		return tlsConfig, fmt.Errorf("failed to read kotsadm-tls secret from cluster: %w", err)
```

Comment:
i think we should try to make error not stutter. e.g. "failed to...: failed to: ..."

```suggestion
		return tlsConfig, fmt.Errorf("read kotsadm-tls secret from cluster: %w", err)
```